File: Go/919-complete-binary-tree-inserter.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 type CBTInserter struct {
    list [] *TreeNode
}


func Constructor(root *TreeNode) CBTInserter {
    var TreeList CBTInserter
    var queue []*TreeNode
    queue = append(queue, root)
    for len(queue) != 0 {
        cur := queue[0]
        queue = queue[1:]
        if cur.Left != nil {
            queue = append(queue, cur.Left)
        }
        if cur.Right != nil {
            queue = append(queue, cur.Right)
        }
        TreeList.list = append(TreeList.list, cur)
    }
    return TreeList
}


func (this *CBTInserter) Insert(v int) int {
    var cur TreeNode
    cur.Val = v
    father := this.list[(len(this.list) - 1) / 2]
    if father.Left == nil {
        father.Left = &cur
    } else {
        father.Right = &cur
    }
    this.list = append(this.list, &cur)
    return father.Val
}


func (this *CBTInserter) Get_root() *TreeNode {
    return this.list[0]
}


/**
 * Your CBTInserter object will be instantiated and called as such:
 * obj := Constructor(root);
 * param_1 := obj.Insert(v);
 * param_2 := obj.Get_root();
 */

File: Python/919-complete-binary-tree-inserter.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class CBTInserter:

    def __init__(self, root: TreeNode):
        self.TreeList = [root]
        for node in self.TreeList:
            if node.left:
                self.TreeList.append(node.left)
            if node.right:
                self.TreeList.append(node.right)
        
    def insert(self, v: int) -> int:
        father, cur = self.TreeList[(len(self.TreeList) - 1) // 2], TreeNode(v)
        if father.left:
            father.right = cur
        else:
            father.left = cur
        self.TreeList.append(cur)
        return father.val

    def get_root(self) -> TreeNode:
        return self.TreeList[0]


# Your CBTInserter object will be instantiated and called as such:
# obj = CBTInserter(root)
# param_1 = obj.insert(v)
# param_2 = obj.get_root()